diff --git a/dist/src/socket-to-conn.js b/dist/src/socket-to-conn.js
index 2ef149718745e5901cdce459d73dcb3d4b7e596e..a8a2ba0830c4f1cb42bd555112ee01dead908c2f 100644
--- a/dist/src/socket-to-conn.js
+++ b/dist/src/socket-to-conn.js
@@ -88,6 +88,10 @@ class TCPSocketMultiaddrConnection extends AbstractMultiaddrConnection {
         await pEvent(this.socket, 'close', options);
     }
     sendReset() {
+        if (this.socket.writableEnded) {
+            this.socket.destroy();
+            return;
+        }
         this.socket.resetAndDestroy();
     }
     sendPause() {
diff --git a/src/socket-to-conn.ts b/src/socket-to-conn.ts
index c4a085b17b6326c96b305741af6b1717cd8a2714..4fba3375008e899bb60eef0382cb4d9c6245525f 100644
--- a/src/socket-to-conn.ts
+++ b/src/socket-to-conn.ts
@@ -123,6 +123,11 @@ class TCPSocketMultiaddrConnection extends AbstractMultiaddrConnection {
   }
 
   sendReset (): void {
+    if (this.socket.writableEnded) {
+      this.socket.destroy()
+      return
+    }
+
     this.socket.resetAndDestroy()
   }
 
